### ISSUE TITLE
High rate of hash collisions in RuntimeBlockStateRegistry

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#5604]
## Description
Most blocks will have few states, but since a small selection have a large number (notably walls), the least significant bits of internal blockstate IDs will often be zero or other small numbers.

This is problematic because these bits are used for hash keys in arrays, meaning that the majority of blocks will end up in the same collision list. This results in a loss of performance when fetching blocks from RuntimeBlockStateRegistry.

How much of an impact this has on overall performance is not clear, but if past examples are anything to judge by it's not good.

## Possible solutions

Addressing this seems fairly simple: put the type ID in the low bits of the state ID, instead of the state data. This would hand off the reduced performance to blocks with lots of states instead, such as walls, while regular blocks would perform normally.

Another possible option is to morton2d_encode(typeID, stateData), although this would still slightly affect performance for the common cases.

Alternatively, a 2D array could be used for blockstate indexing, which would be equally fast for both cases, but probably slightly slower than the other options.

**Comment by dktapps** [comment]
`array_load_factor` from `ext-arraydebug` shows a load factor of `1.95`, which is not ideal, but not as bad as it could be.
`array_hash_distribution` shows that one hash key ends up with 22 elements under it.

**Comment by dktapps** [comment]
This also negatively affects the performance of block property caches in PM5, such as those for blast resistance and light level / light opacity levels.

**Comment by dktapps** [comment]
I experimented with morton2d for this, but this evenly reduces performance for all blocks, since it halves the number of distinct bits available to blocks without any states, effectively distributing the problem to all blocks, but unfairly towards stateless blocks.

The 2D array solution might end up being the best solution for this, but it's not ideal, due to the increase in either memory footprint or branching complexity in fast paths.

Another possible option is to assign an opaque incrementing ID to each known state a la Minecraft network IDs, which would linearize the caches. The problem with this is that we would either need to share these opaque IDs with other threads, or we would need to have `FastChunkSerializer` encode blocks using some persistent format when transferring chunks between threads. This is doable, but would reduce the performance of `FastChunkSerializer` and make it less able to be implemented natively.

**Comment by dktapps** [comment]
Looks like xxhash64 might be a workable solution for this, although it'll need a native implementation that directly accepts and returns an `int`, to avoid unnecessary string conversions.

**Comment by dktapps** [comment]
XORing the state data with the type ID improves the lower bits distribution quite a bit, reducing the worst-case number of collisions from 11 to 5, and improving overall load factor from 1.6 to 1.3.

Use of xxhash(stateID) didn't significantly improve the distribution with any version (I got it as low as 1.24).

Use of xxhash(typeID << 8 | ((stateData ^ xxhash(typeID) & 0xff)) improves the load factor to as low as 1.19 with a worst-case of 3 collisions (using xxhash3_64bits_int).

TL;DR: I don't think we'll get much more improvement than what we have with ffe3556be1c8c20218190da184510518694f3234.
